mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -283,10 +281,7 @@ private boolean updateAvailableConnection(Connection c, ConnectionException fail
             mConnectionFailureReason = null;
             mAvailableConnection = c;
         }
-        WebImageCache imageCache = MyWebImage.getWebImageCache();
-        if (imageCache != null) {
-            imageCache.clear();
-        }
+        CacheManager.getInstance(ctx).clearCache();
```

Comment:
The icon cache doen't seem to be persistent across app close and open now.

Reply:
Hmm, I guess we either
- need to look into the cache headers sent by the server or
- drop the HTTP cache and do disk caching of bitmaps ourselves (or use a library for it)

I'll see what headers the server sends.

Reply:
Should be OK now. I removed the clear-cache-on-connection-change logic and fixed up icon caching to not do HTTP requests if there's a cache entry.

I'll have a look why I introduced this cache clearing logic in the first place (https://github.com/openhab/openhab-android/pull/562/commits/1b2f880776b6d2c647ee0b261c883d050b83850a), as I don't remember anymore :-/ ... it was related to switching between my OH1 server and the (OH2) demo server, after which I had mixed OH1 and OH2 icons, but that shouldn't have happened anyway as the URL is used as cache key ... if that switching scenario doesn't reproduce the problem, I think we can safely stop clearing the cache.

Reply:
It seems that the icon cache is still cleared when restarting the app.

Reply:
How do you tell? It looks fine to me now, in that icons show up instantly.

Reply:
I overlooked this when I searched for open issues on this PR.

Reply:
Yes, they're loaded again from the perspective of WidgetImageView, but they're fetched from the OkHttp cache. To verify this, I applied this patch:

```diff
diff --git a/mobile/build.gradle b/mobile/build.gradle
index 40c1e197..76e9afc8 100644
--- a/mobile/build.gradle
+++ b/mobile/build.gradle
@@ -102,6 +102,7 @@ dependencies {
     implementation 'com.android.support.constraint:constraint-layout:1.1.0'
     implementation 'org.jmdns:jmdns:3.5.3'
     implementation 'com.squareup.okhttp3:okhttp:3.9.1'
+    implementation 'com.squareup.okhttp3:logging-interceptor:3.9.1'
     implementation 'com.github.BigBadaboom:androidsvg:3511e136498da94018ef9fa438895984ea9b99db'
     implementation 'com.github.apl-devs:appintro:v4.2.2'
     // Google Maps
diff --git a/mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java b/mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java
index 2a2c4d27..6e8e40ca 100644
--- a/mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java
+++ b/mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java
@@ -27,6 +27,7 @@ import java.security.cert.X509Certificate;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
@@ -48,6 +49,8 @@ import okhttp3.MediaType;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
 import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.logging.HttpLoggingInterceptor;
 
 public abstract class HttpClient {
     private static final String TAG = HttpClient.class.getSimpleName();
@@ -77,6 +80,15 @@ public abstract class HttpClient {
     protected HttpClient(Context context, SharedPreferences prefs, String baseUrl) {
         final Context appContext = context.getApplicationContext();
         mClient = new OkHttpClient.Builder()
+                .addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC))
+                .addInterceptor(chain -> {
+                    Response response = chain.proceed(chain.request());
+                    Log.d(TAG, String.format(Locale.US, "For %s: network return code %d, cache %d",
+                            response.request().url().toString(),
+                            response.networkResponse() != null ? response.networkResponse().code() : -1,
+                            response.cacheResponse() != null ? response.cacheResponse().code() : -1));
+                    return response;
+                })
                 .cache(CacheManager.getInstance(context).getHttpCache())
                 .build();
         mBaseUrl = baseUrl != null ? HttpUrl.parse(baseUrl) : null;
```

Output for `adb logcat -s HttpClient:V WidgetImageView:V` for the widget overview page after clearing image cache:

```
06-01 08:54:36.824 21907 22020 D HttpClient: For https://demo.openhab.org:8443/rest/sitemaps/demo/0202: network return code 200, cache -1
06-01 08:54:36.864 21907 21907 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=1&format=PNG
06-01 08:54:36.871 21907 21907 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=1&format=PNG
06-01 08:54:36.903 21907 21988 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=1&format=PNG: network return code 200, cache -1
06-01 08:54:36.904 21907 22019 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=1&format=PNG: network return code 200, cache -1
06-01 08:54:38.639 21907 22019 D HttpClient: For https://demo.openhab.org:8443/rest/items/Scene_General: network return code 200, cache -1
06-01 08:54:38.799 21907 22020 D HttpClient: For https://demo.openhab.org:8443/rest/sitemaps/demo/0202: network return code 200, cache -1
06-01 08:54:38.825 21907 21907 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG
06-01 08:54:38.829 21907 21907 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG
06-01 08:54:38.864 21907 21988 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG: network return code 200, cache -1
06-01 08:54:38.867 21907 22019 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG: network return code 200, cache -1
06-01 08:54:40.975 21907 22019 D HttpClient: For https://demo.openhab.org:8443/rest/items/Scene_General: network return code 200, cache -1
06-01 08:54:41.293 21907 22020 D HttpClient: For https://demo.openhab.org:8443/rest/sitemaps/demo/0202: network return code 200, cache -1
```

All images are actually fetched, as expected. After closing and reopening the app:

```
06-01 09:45:05.163 23072 23122 D HttpClient: For https://demo.openhab.org:8443/rest/sitemaps/demo/0202: network return code 200, cache -1
06-01 09:45:05.230 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/switch?state=ON&format=PNG
06-01 09:45:05.234 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/switch?state=ON&format=PNG: network return code -1, cache 200
06-01 09:45:05.240 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/switch?state=ON&format=PNG
06-01 09:45:05.244 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/switch?state=ON&format=PNG: network return code -1, cache 200
06-01 09:45:05.264 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG
06-01 09:45:05.268 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG: network return code -1, cache 200
06-01 09:45:05.274 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG
06-01 09:45:05.278 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/sofa?state=3&format=PNG: network return code -1, cache 200
06-01 09:45:05.289 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/temperature?state=17&format=PNG
06-01 09:45:05.293 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/temperature?state=17&format=PNG: network return code -1, cache 200
06-01 09:45:05.308 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/slider?state=100&format=PNG
06-01 09:45:05.314 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/slider?state=100&format=PNG: network return code -1, cache 200
06-01 09:45:05.323 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/slider?state=%23a82c8d&format=PNG
06-01 09:45:05.327 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/slider?state=%23a82c8d&format=PNG: network return code -1, cache 200
06-01 09:45:05.335 23072 23072 I WidgetImageView: Refreshing image at https://demo.openhab.org:8443/icon/rollershutter?state=100&format=PNG
06-01 09:45:05.339 23072 23131 D HttpClient: For https://demo.openhab.org:8443/icon/rollershutter?state=100&format=PNG: network return code -1, cache 200
```

All images are now loaded from local OkHttp cache, no HTTP request is actually made.

(BTW, I've been thinking of integrating that debug logging patch for some time, but I'm wondering about the best way to enable/disable it)

Reply:
Got it :) You could add it and only log when the debug mode in the settings is enabled.